research/minigo/dualnet_model.py

```diff
@@ -0,0 +1,236 @@
+    """
+    initial_output = _conv_block(
+        inputs=inputs, filters=self.num_filters,
+        kernel_size=self.kernel_size, training=training)
+    # the shared stack
+    shared_output = initial_output
+    for _ in range(self.num_shared_layers):
+      shared_output = _res_block(
+          inputs=shared_output, filters=self.num_filters,
+          kernel_size=self.kernel_size, training=training)
+
+    # policy head
+    policy_conv2d = _conv2d(inputs=shared_output, filters=2, kernel_size=[1, 1])
+    policy_batchn = _batch_norm(inputs=policy_conv2d, training=training,
+                                center=False, scale=False)
+    policy_relu = tf.nn.relu(policy_batchn)
+    policy_logits = tf.layers.dense(
+        tf.reshape(policy_relu, [-1, self.go_size * self.go_size * 2]),
+        self.go_size * self.go_size + 1)
+
+    # value head
+    value_conv2d = _conv2d(shared_output, filters=1, kernel_size=[1, 1])
+    value_batchn = _batch_norm(value_conv2d, training,
+                               center=False, scale=False)
+    value_relu = tf.nn.relu(value_batchn)
+    value_fc_hidden = tf.nn.relu(tf.layers.dense(
+        tf.reshape(value_relu, [-1, self.go_size * self.go_size]),
+        self.fc_width))
+    value_logits = tf.reshape(tf.layers.dense(value_fc_hidden, 1), [-1])
+
+    return policy_logits, value_logits
+
+
+def model_fn(features, labels, mode, params, config=None):  # pylint: disable=unused-argument
+  """DualNet model function.
+
+  Args:
+    features: tensor with shape
+      [BATCH_SIZE, self.go_size, self.go_size, features.NEW_FEATURES_PLANES]
+    labels: dict from string to tensor with shape
+      'pi_tensor': [BATCH_SIZE, self.go_size * self.go_size + 1]
+      'value_tensor': [BATCH_SIZE]
+    mode: a tf.estimator.ModeKeys (batchnorm params update for TRAIN only)
+    params: a dict of hyperparams
+    config: ignored; is required by Estimator API.
+  Returns:
+    EstimatorSpec parameterized according to the input params and the current
+    mode.
+  """
+  model = Model(params.num_filters, params.num_shared_layers, params.fc_width,
+                params.go_size)
+  policy_logits, value_logits = model(
+      features, mode == tf.estimator.ModeKeys.TRAIN)
+
+  policy_output = tf.nn.softmax(policy_logits, name='policy_output')
+  value_output = tf.nn.tanh(value_logits, name='value_output')
+
+  # Calculate model loss (combined_cost)
+  policy_cost = tf.reduce_mean(
+      tf.nn.softmax_cross_entropy_with_logits(
+          logits=policy_logits, labels=labels['pi_tensor']))
+  value_cost = tf.reduce_mean(
+      tf.square(value_output - labels['value_tensor']))
+  l2_cost = params.l2_strength * tf.add_n(
+      [tf.nn.l2_loss(v) for v in tf.trainable_variables()
+       if 'bias' not in v.name])
+  combined_cost = policy_cost + value_cost + l2_cost
+
+  # For logging purpose
+  policy_entropy = -tf.reduce_mean(tf.reduce_sum(
+      policy_output * tf.log(policy_output), axis=1))
+
```

Comment:
Throughout this fn, it would be helpful to have "walk through" explanations for those who are not familiar-- what are each of the losses used for? Why do we care about policy entropy? Etc.